add a compiler flag to use int64 as tensor size

apeforest opened the pull request:
## Description ##
This PR is to fix the performance degradation reported in https://github.com/apache/incubator-mxnet/issues/14496.

The performance degradation is introduced by PR https://github.com/apache/incubator-mxnet/pull/11742. I have verified the performance degradation in one of the operators transpose using a script below:

```
import mxnet as mx
import time
import numpy as np

sizes = [10, 50, 100,200,500]
iters = [10000,1000,500,200,20]
times = []
for size in range(len(sizes)):
    data = []
    s = sizes[size]
    print(s)
    for i in range(iters[size]):
        x = mx.nd.ones((s,s,s))
        mx.nd.waitall()
        start = time.time()
        y = mx.nd.transpose(x,(2,0,1))
        mx.nd.waitall()
        data.append((time.time() - start)*1000)
        #print(data[-1])                                                                                                                                                                            
    times.append(data)

print('mxnet version: %s' % mx.__version__)
for s in range(len(sizes)):
    print('--------------------')
    print('size: %s' % str(sizes[s]))
    print('p50: %4.2f ms' % np.percentile(times[s],50))
    print('p90: %4.2f ms' % np.percentile(times[s],90))
    print('p99: %4.2f ms' % np.percentile(times[s],99))
```

By changing the `index_t` type from `int64_t` to `int32_t` can consistently change the 50-percentile runtime of tranpose of size 500 from 5000ms to 2400ms. 

By changing the data type in the operator (https://github.com/dmlc/mshadow/blob/master/mshadow/extension/transpose.h#L70) alone, can also reduce the 50-percentile runtime of size 500 to 2600ms.

I thereforce come to the conclusion that the performance degradation is caused by the runtime of  integer arithmetic operations between 32-bit integers and 64-bit integers.

To further experiment with the arithmetic operations alone, I tested using a small program [here](https://github.com/apeforest/doraemon/blob/master/perf32vs64.cc). The runtime results are shown below:

```
result = 49995000
Add 32-bit time in ms 1359
result = 49995000
Add 64-bit time in ms 1971
result = 349965000
Add Mul 32-bit time in ms 1196
result = 349965000
Add Mul 64-bit time in ms 3477
result = 7137858
Add Div 32-bit time in ms 2878
result = 7137858
Add Div 64-bit time in ms 8499
```

I can think of three solutions to this problem:
(1) Add a compilation flag to choose data types for tensor size (This PR)
(2) Add an environment variable to choose data type for tensor size at runtime
(3) Choose data type for tensor size at runtime based on the size of the tensor

Given the expression template used in mshadow for operators, either (2) or (3) requries a significant change in the mshadow library. (1) can be used as a quick solution to fix performance degradation reported in several issues: https://github.com/apache/incubator-mxnet/issues/14496, https://github.com/apache/incubator-mxnet/issues/13928, https://github.com/apache/incubator-mxnet/issues/14563, https://github.com/apache/incubator-mxnet/issues/14569

Any other suggestion is appreciated.

This PR also depends on https://github.com/dmlc/mshadow/pull/371

## Checklist ##
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

apeforest commented:
@pengzhao-intel @TaoLv Please help to review. Thanks!

abhinavs95 commented:
@mxnet-label-bot add [pr-work-in-progress, backend]

TaoLv commented:
Do we need to change CI to cover this new flag? Do we need to add it to the runtime feature set? @larroy

wkcn commented:
Thank you for the quick fix!

I think we can use the solution (3): 'Choose data type for tensor size at runtime based on the size of the tensor' in the performance bottleneck firstly.

pengzhao-intel commented:
Agree with adding the new flag to switch to int64 when the user explicitly know they want to use the large array, like DGL. For the normal case, int32 is enough w/o performance issue.

@TaoLv do we need to switch off MKLDNN when int64 is enabled before MKL-DNN 1.0?

@wkcn would you mind mentioning the advantage of 3)?

wkcn commented:
@pengzhao-intel Yes. For users, it is confusing that there will be two versions of mxnet: mxnet and mxnet-largetensor.  In other words, users need to rebuild the source, when they use large tensor. It is too complex.

```c++
```

apeforest commented:
@wkcn Thanks for the suggestion. This macro is good for certain simple operators but I found it difficult to apply to all operators given the different architecture of operator implementation. E.g. transpose operator is currently using the expression template pattern in mshadow. It seems a big surgery to apply this macro to operators in mshadow. Besides, int64_t is used as type of data members in the tensor datastructure. This approach will not directly change the internal methods in tensor or requires static cast. Any other thoughts on this?

samskalicky commented on .gitmodules:
```diff
@@ -1,6 +1,6 @@
 [submodule "3rdparty/mshadow"]
 	path = 3rdparty/mshadow
-	url = https://github.com/dmlc/mshadow.git
+	url = https://github.com/apeforest/mshadow.git
```
I see why you did this for testing, but do we want to check this is to master and depend on your fork?

apeforest replied:
Yes, after mshadow PR is merged, I will update this here accordingly.

eric-haibin-lin replied:
https://github.com/dmlc/mshadow/pull/371 has been merged

samskalicky commented on make/config.mk:
```diff
@@ -214,6 +214,10 @@ EXTRA_OPERATORS =
 USE_CPP_PACKAGE = 0
 
+# Use int64_t type for tensor size.
+# This will cause performance degradation reported in issue #14496
```
Maybe add something to say when to use it too, like "Set to 1 for large tensor sizes, when tensors will have large shapes such as for graphs" or something...?

apeforest replied:
updated.

samskalicky commented on make/config.mk:
```diff
@@ -216,6 +216,7 @@ USE_CPP_PACKAGE = 0
 
 # Use int64_t type for tensor size.
 # This will cause performance degradation reported in issue #14496
+# Set to 1 for large tensor size greater than 2 billion
```
not to be too picky, but what does this mean "tensor size greater than 2 billion"? Does this mean that if you have more than 2 billion elements in a single dimension you need this? or 2 billion elements in the whole tensor?

apeforest replied:
updated

larroy commented on include/mxnet/tensor_blob.h:
```diff
@@ -442,7 +442,7 @@ class FieldEntry<mxnet::TShape>
     }
     if (enforce_nonzero_) {
-      for (mxnet::index_t i = 0; i < v.ndim(); ++i) {
+      for (int i = 0; i < v.ndim(); ++i) {
```
why int here?

apeforest replied:
dimensions will not be hight and `int` is used across many template for dim such as https://github.com/dmlc/mshadow/blob/95ebe0f109ae021d0d66e2a627ccfc55c3253b55/mshadow/tensor.h#L43

larroy commented on include/mxnet/tensor_blob.h:
```diff
@@ -456,7 +456,7 @@ class FieldEntry<mxnet::TShape>
     return this->self();
   }
-  inline FieldEntry<mxnet::TShape> &set_expect_ndim(mxnet::index_t ndim) {
+  inline FieldEntry<mxnet::TShape> &set_expect_ndim(int ndim) {
```
why int? I know dimensions can't be too high but wouldn't it make sense to have a typedef for it?

apeforest replied:
Prefer `int` to defined type for better readability. `unsigned int` is discouraged to use as data type. See https://google.github.io/styleguide/cppguide.html#Integer_Types

larroy replied:
I get your point. I'm not convinced that the dimensions type should be signed, as it makes no sense. Agree that the int width should be enough. I personally would make it unsigned.

larroy replied:
Thanks @reminisce  and @apeforest . Let's resolve this comment then.

larroy replied:
Adding to this, the type is signed ok I understand that. I still think we should specify the width of the type, say int32_t

anirudh2290 commented on Makefile:
```diff
@@ -188,6 +188,11 @@ ifeq ($(USE_OPERATOR_TUNING), 1)
 endif
 
+ifeq ($(USE_INT64_TENSOR_SIZE), 1)
+   CFLAGS += -DMSHADOW_INT64_TENSOR_SIZE=1
```
Since USE_INT64_TENSOR_SIZE has been set by default to 0 this is a breaking change. Are we going to add an exception for semver for the next minor release ?

apeforest replied:
This does not break API but will impact cases for large tensors. How should we handle this cases?

anirudh2290 replied:
Ya probably doesn't fall under semver. We should still let users like dgl know about the change and how they can build for large tensor support.

yuxihu replied:
Will the MXNet pip packages be built with large tensor support disabled after this PR is merged? If so, for users who want large tensor support, do they have to build MXNet from source?

apeforest replied:
Yes, users who need large tensor support will have to build MXNet from source.

yuxihu replied:
I think we need to document this. Maybe on the build from source page?

apeforest replied:
It's documented in the config.mk file: https://github.com/apache/incubator-mxnet/pull/14570/files#diff-ea86416d0f764a60d29337663f7a1767R220

samskalicky commented:
@larroy the dimension can be -1 if its unknown, or if -1 refers to the last dimension. So we do have to use signed types.

eric-haibin-lin commented on include/mxnet/tensor_blob.h:
```diff
@@ -223,11 +223,11 @@ class TBlob {
    * \return the size
    */
-  inline index_t size(index_t idx) const {
+  inline size_t size(index_t idx) const {
```
This will break shapes with negative numbers? @reminisce

reminisce replied:
As another effort in parallel, we need to use `-1` to represent unknown dim size to support zero-size tensors. So the type has to be kept as a signed integer.

apeforest replied:
updated

larroy commented:
@reminisce why aren't we using ndim=1 for scalars?

reminisce commented:
@larroy Tensors with `ndim=1` are not scalars, but 1D tensors, aka vectors.

apeforest commented:
Added build with this new compiler flag on to CI and added a nightly test for large tensor per @TaoLv suggestion. @larroy @marcoabreu Please help to review.

apeforest commented:
@reminisce replaced nnvm::Tuple with mxnet::Tuple as we discussed in dev list. Please help to review. Thanks!

larroy commented:
Thanks a lot for the detailed report and analysis @apeforest and @samskalicky . From the data you guys have provided I'm missing a disassembly dump of the small loop that is supposed to be slow. Could you guys provide this with "objdump -d" or similar? I find it surprising that the degradation is only due to data widening. I suspect the cause is more the memory access than the wider arithmetic operation itself. I think having the additional data point of the assembly would help reach a more solid conclusion.

larroy commented:
To be honest, while I understand the rationale of the solution proposed. I think it would be better if we could avoid having yet another build flavour if possible.  Could we for example, narrow the computation at runtime in case the number of dimensions doesn't overflow? Also see my previous comment above to see if we can work with memory access to make it almost as fast with the wider type in this particular computation. If I understood correctly the regression is only in transposing? the other operators are fine? We could try to find a solution only for this operator then instead of adding a new build flavour.

samskalicky commented:
@larroy we dont have any data that guarantees only the transpose operator is affected. In reality it could be all mshadow implemented operators that are affected. We dont have performance tests for each operator that we could run and verify. There are efforts to look into putting together some operator performance tests, but i dont think we should wait on this fix for that. 

This PR is important to restore "balance to the force" until we can work on a better solution. Since new functionality for large tensors was already added and is being used for the Deep Graph Library effort we need to provide a way for them to use the large tensor feature. At the same time, we need to fix the performance degradation. So this is fix (adding the flag) the common middle ground between the two.

larroy commented:
@samskalicky @apeforest makes sense. I just wanted to raise this point of view. If you guys think this is a tradeof we should do now at the cost of adding an additional build flavour and we revisit later once we have per operator performance metrics, I'm fine with that.

larroy commented on ci/docker/runtime_functions.sh:
```diff
@@ -755,6 +755,35 @@ build_ubuntu_gpu_cmake() {
+build_ubuntu_gpu_large_tensor() {
+    set -ex
+    build_ccache_wrappers
+    make  \
```
why not cmake?

apeforest replied:
I saw both `make` and `cmake` used in runtime_functions.sh such as https://github.com/apache/incubator-mxnet/blob/master/ci/docker/runtime_functions.sh#L358

Sholud we only use `cmake` in the future?

larroy replied:
We have all agreed to use CMake, so we should not introduce new make builds unless there's a problem with CMake, this is my understanding.

apeforest replied:
Sure, updated the process to cmake. I think we may need a seperate PR to update all runtime_functions to cmake otherwise it may cause confusion to future developers.

larroy commented on ci/docker/runtime_functions.sh:
```diff
@@ -755,6 +755,35 @@ build_ubuntu_gpu_cmake() {
+build_ubuntu_cpu_large_tensor() {
+    set -ex
+    build_ccache_wrappers
+    export CC="ccache gcc"
```
build_ccache_wrappers should be enough. My PR also fixes that function: https://github.com/apache/incubator-mxnet/pull/14631

Would be good to merge then no need to export the CC and CXX variables.

apeforest replied:
No longer needed after I change to using cmake

larroy commented on CMakeLists.txt:
```diff
@@ -295,6 +296,13 @@ else()
   add_definitions(-DMXNET_USE_NCCL=0)
 endif()
 
+if (USE_INT64_TENSOR_SIZE)
```
Can you also add this to the runtime feature detection code please?

https://mxnet.incubator.apache.org/api/python/libinfo/libinfo.html

apeforest replied:
Added.

apeforest commented:
@TaoLv Per @marcoabreu suggestion, I moved the tests from CI to nightly to ease the load of CI pipeline. Please let me know if this is okay with you. Thanks!

TaoLv commented:
@apeforest Thank you. That sounds reasonable. But I remember I couldn't find the nightly test report for large tensor before. Do you know where we can read it once this PR is merged?

apeforest commented:
> @apeforest Thank you. That sounds reasonable. But I remember I couldn't find the nightly test report for large tensor before. Do you know where we can read it once this PR is merged?

Tao, I think the nightly tests are here: http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/NightlyTests/activity

@marcoabreu  Could you please confirm? Thanks!

apeforest commented:
@TaoLv Do you have any more concerns for this PR? Thanks!

apeforest commented:
> Are we checking for narrowing / overflow since we have this additional wider type for dimensions?

This PR itself is not narrowing/widening the data type for dimension. It simply adds a compiler flag to control this.

apeforest commented:
@eric-haibin-lin @yuxihu @samskalicky @anirudh2290 @reminisce  Please let me know if your concerns have been addressed. Thanks!

marcoabreu commented:
LGTM from a CI perspective

TaoLv commented:
Thank you for the fix, @apeforest. Now it's approved. Just copy some comments from the mshadow PR here: int64 data type only affects the definition of some variables and the total size of a NDArray. We won't pass any NDArray with dim[x] > INT32_MAX into a operator. So it will not break the existing definition of BLAS and MKL-DNN APIs.

apeforest commented:
@marcoabreu @TaoLv @anirudh2290 could you please help to merge it if it’s good to ship? Thanks

edisongustavo commented on tests/nightly/JenkinsfileForBinaries:
```diff
@@ -59,6 +78,22 @@ core_logic: {
+    'Test Large Tensor Size: CPU': {
+      node(NODE_LINUX_CPU) {
+        ws('workspace/large_tensor-cpu') {
+            utils.unpack_and_init('cpu_int64', mx_cmake_lib)
```
I guess this should have been `ubuntu_cpu_int64`.

The nightly tests are failing because no one is publishing this stash: http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/NightlyTestsForBinaries/detail/master/292/pipeline

As you can see from "Build / CPU: USE_INT64_TENSOR_SIZE", it publishes `ubuntu_cpu_int64`.
